[Modal Layout Picker] Create Modal window for Modal Layout Picker

antonis opened the pull request:
**Fixes:** https://github.com/wordpress-mobile/gutenberg-mobile/issues/2419

**Related PRs:**
- `gutenberg-mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/2494
- `gutenberg`: https://github.com/WordPress/gutenberg/pull/24061

## Description
Adds feature flags and a container view for the Modal Layout Picker

## To test:

### Classic Editor and Gutenberg Editor with Modal Layout Picker Enabled.
- Create a new page from My Site
    - Create a new page by clicking on the Floating Toolbar Button on the My Site Page then select "Site Page"
    - Expect to see the modal layout picker container
    - Select "Create Blank Page"
    - Expect to see the page editor
- Create a new page from Pages.
    - Create a new page by clicking on the fab icon on the Pages screen.
    - Expect to see the modal layout picker container
    - Select "Create Blank Page"
    - Expect to see the page editor
- Create a new page from Pages when the site has no pages.
    - Create a new page by clicking on the 'Create a page' button.
    - Expect to see the modal layout picker container
    - Select "Create Blank Page"
    - Expect to see the page editor

### On the Modal Layout Picker
- Selecting the back icon at the top left should close the modal without creating a page.
- Swiping down around the "Grip" in the top middle of the modal should close the modal without creating a page.
- Scrolling up in the stubbed rows should collapse the header
- Toggle Dark mode and check the style seems consistent with other views

## Screenshots:
| Container | Action |
|---|---|
|![device-2020-07-22-115223](https://user-images.githubusercontent.com/304044/88156486-2c201880-cc12-11ea-9966-ed88ae1553e0.png)|![device-2020-07-22-114828](https://user-images.githubusercontent.com/304044/88156395-072ba580-cc12-11ea-8361-849d96f2b868.gif)|
|---|---|
| Dark Mode | Collapsed |
|---|---|
|![device-2020-07-22-115242](https://user-images.githubusercontent.com/304044/88156554-42c66f80-cc12-11ea-8670-fb90da1e6ca9.png)|![device-2020-07-22-115310](https://user-images.githubusercontent.com/304044/88156561-44903300-cc12-11ea-87bc-b613a8941ccf.png)|


PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

antonis commented:
After the [Consolidation of the Gutenberg initialProps handling](https://github.com/wordpress-mobile/WordPress-Android/pull/12444) a refactoring is needed in the way the MLP configuration is passed

antonis commented:
Hello @iamthomasbishop 👋 ,

I've completed a first version of the MLP container. Please let me know if you have any design feedback on the attached screenshots or the the APK [here](https://62430-9306568-gh.circle-artifacts.com/0/Artifacts/WordPress-pr-12482-build-62430.apk).
I tried to use the existing text styles of the App. Please provide more information on this to fully align with the provided designs.
Thank you in advance :)

mchowning commented:
> This has been updated with the Consolidation of the Gutenberg initialProps handling and should not be merged till gutenberg/after_release_1.33.0 is merged

:wave: @antonis ! I think if you retarget this PR to the `gutenberg/after_release_1.33.0` branch that would both make the diff on this PR a lot smaller and it would allow you to merge this PR before `gutenberg/after_release_1.33.0` gets merged to `develop`. If you did that and `gutenberg/after_release_1.33.0` got merged before this PR was merged you could just retarget it back to `develop` again (I think Github actually might do that for you automatically when the branch you are targeting gets merged). This certainly isn't critical, it's just a suggestion.

antonis commented:
> I think if you retarget this PR to the gutenberg/after_release_1.33.0 branch that would both make the diff on this PR a lot smaller and it would allow you to merge this PR before gutenberg/after_release_1.33.0 gets merged to develop.

That sound good. I made the change.
Thank you for the prompt feedback @mchowning :)

iamthomasbishop commented:
Hey @antonis 👋 Took the build for a spin, and it's coming along nicely. Not seeing any big issues, but here are my notes.

## Sheet surface

I'm not sure if this is intentional or not, but I noticed when you're scrolled to the top of the view and pull down softly, we have a snap point where the sheet gets docked at ~2/3 height ([video example](https://cloudup.com/c5q9nhX1_9g)). We probably only need 2 heights — full-screen and closed.

I'm not sure if this something we can control, but when you scroll to the bottom of the sheet and get the usual "ripple" effect (the blue "wave" that shows when you scroll to the end of a container), there looks to be a ~16dp gap between that and the top edge of the footer bar.  ([video example](https://cloudup.com/cgaFPF-HVTR))

## Collapsed app bar style

The text label on the app bar when it's in a collapsed state doesn't look centered. ([screenshot](https://cloudup.com/cxjMJ3ku6_S))

## Title/subtitle

The sizing of the title and subtitle look a little small — I believe the title should be set in Headline 4 and the subtitle in Body 1, but it's hard to tell what the current values are in this build — would you mind confirming?

Is there any way we can use the system-default serif font (Noto Serif)? My original mockups used the **bold** weight because that's what we were leaning towards using for headlines across the app, but I'll have to check in to see if that's the final decision.

The spacing between the app bar and title also look a little tight — can you confirm how much space is between these (I believe it should be 16, although IIRC the line-height of the labels might be slightly different, which would result in a slight diff)? 

The spacing between the title and subtitle look a bit too much, but this might be an optical illusion bc of the too-small spacing mentioned above (this also should be 16).

[https://cloudup.com/cnO1Q2EZM9c](https://cloudup.com/cnO1Q2EZM9c)

## Status bar color

Would it be possible to adapt the status bar to a light variation  (white background, dark text) when the sheet is fully open? For reference, we're also using this style on [post/page preview](https://cloudup.com/cE2UZd5lHHP).

iamthomasbishop commented:
@antonis an update regarding my comment on font-weight of the headings:

> Is there any way we can use the system-default serif font (Noto Serif)? My original mockups used the bold weight because that's what we were leaning towards using for headlines across the app, but I'll have to check in to see if that's the final decision.

We are using the bold weight of Noto Serif for headings as I thought, so we can use that for this sheet 😄

chipsnyder commented on WordPress/build.gradle:
```diff
@@ -109,11 +111,13 @@ android {
         wasabi { // "hot" version, can be installed along release, alpha or beta versions
             applicationId "org.wordpress.android.beta"
             dimension "buildType"
+            buildConfigField "boolean", "MODAL_LAYOUT_PICKER", "true"
```
Hey @antonis, was coming back to this to check on the progress. This is something I didn't notice while scoping out the issue, but I wonder if this should be enabled under `zalpha` instead of under the `wasabi` and `jalapeno` based on the comment `// alpha version - enable experimental features`

I didn't do a full review yet but wanted to call out we might want to check with someone from the main WPAndroid teams to double-check.

antonis replied:
Hello @chipsnyder 👋 ,
I initially enabled this for [`vanilla` and `jalapeno`](https://github.com/wordpress-mobile/gutenberg-mobile/issues/2419) but after talking to @pinarol I changed it to `wasabi` and `jalapeno` since we did not want this enabled in beta yet. I cross checked this configuration with @malinajirka last week.
To my understanding we can enable this in `zalpha` too.

malinajirka replied:
Just to summarise. It really depends on who should have access to this feature.

- `wasabi` is for developers
- `jalapeno` is for reviewers (automatic CI builds on PRs - needs to be manually downloaded and installed)
- `zalpha` is for alpha testers (closed track distributed through Google Play)
- `vanilla` is for beta testers and users/release (open track + release distributed through Google Play)

chipsnyder replied:
That's more clear now then. Thank you both for the explanation 🙂

antonis commented:
Hello @iamthomasbishop 👋 ,

I updated the PR based on your feedback.

> we have a snap point where the sheet gets docked

removed

> gap between that and the top edge of the footer bar

removed

> The text label on the app bar when it's in a collapsed state doesn't look centered

fixed

> the title should be set in Headline 4 and the subtitle in Body 1

fixed

> We are using the bold weight of Noto Serif for headings as I thought

applied

> The spacing between the app bar and title also look a little tight 

fixed

> The spacing between the title and subtitle look a bit too much

fixed

> adapt the status bar to a light variation (white background, dark text) when the sheet is fully open

Tweaked this to be 'lighter'. I need to investigate further if we can achieve the same result as in post/page preview for this screen.

|A|B|C|D|
|---|---|---|---|
|![device-2020-07-29-114720](https://user-images.githubusercontent.com/304044/88778719-8597c280-d191-11ea-98b1-2f2e2c74b1ec.png)|![device-2020-07-29-114739](https://user-images.githubusercontent.com/304044/88778729-87fa1c80-d191-11ea-90c5-083006b4b79e.png)|![device-2020-07-29-114758](https://user-images.githubusercontent.com/304044/88778731-8892b300-d191-11ea-823c-11a501059546.png)|![device-2020-07-29-114816](https://user-images.githubusercontent.com/304044/88778733-8892b300-d191-11ea-8cf4-1ffa55f20838.png)|

antonis commented:
Also added a sticky header implementation for the categories row as a placeholder to be in sync with [the iOS implementation](https://github.com/wordpress-mobile/WordPress-iOS/pull/14501).

<img width="360" src="https://user-images.githubusercontent.com/304044/88796747-d7991200-d1aa-11ea-9b4f-ecf7364ca9e7.gif">

antonis commented on WordPress/src/main/res/layout/modal_layout_picker_bottom_toolbar.xml:
```diff
@@ -0,0 +1,38 @@
+        android:layout_weight="1"
+        android:text="@string/mlp_create_blank_page" />
+
+    <com.google.android.material.button.MaterialButton
```
The two last buttons are not actually needed now. Just added them for future use and to check the layout behavior

antonis commented on WordPress/src/main/java/org/wordpress/android/ui/mlp/CategoriesItemDecoration.kt:
```diff
@@ -0,0 +1,38 @@
+        scrollingView.draw(canvas)
+        canvas.restore()
+    }
+}
```
This approach on implementing the categories 'sticky' header may be revisited since i has some limitations regarding handling touches. This become apparent while [implementing the categories selection in a next PR](https://github.com/wordpress-mobile/WordPress-Android/pull/12573). A workaround and another approach on this is examined.

chipsnyder commented on WordPress/src/main/java/org/wordpress/android/ui/mlp/ModalLayoutPickerFragment.kt:
```diff
@@ -0,0 +1,143 @@
+/**
+ * Implements the Modal Layout Picker UI based on the [BottomSheetDialogFragment] to inherit the container behavior
+ */
+class ModalLayoutPickerFragment : BottomSheetDialogFragment() {
```
Just as a general naming suggestion but not requiring a change here. `ModalLayoutPickerFragment` feels like it's really inforcing the display mechanism. Because of that on iOS, I went with `GutenbergLayoutPicker__` because that felt like it described more of what it does instead of how it should be displayed plus what it does.

antonis replied:
I agree with you on this Chip. Since the Modal/MLP naming has infected a lot of my classes an resource naming I will leave this refactoring last to avoid any conflicts

antonis commented:
Hello @chipsnyder 👋 ,
thank you for the detailed feedback. I really appreciate this :)
I tackled the reported problems and added landscape mode handling.

![device-2020-08-04-134723](https://user-images.githubusercontent.com/304044/89285490-22a69f80-d659-11ea-97d3-7ececda2910e.png)

chipsnyder commented:
Hey @antonis 👋 

The issue with the large title disappearing seems fixed now good work on that 👍. The transition for the large title feels a lot better but I'd like @iamthomasbishop's opinion on the scroll for that area. 

I'm wondering if it would feel more smooth if we had the large title transition to the small title at the point where the large title was mostly scrolled off the screen instead of at the start of scrolling. That way we prevent a large gap like in this screenshot
<img width=300 src="https://user-images.githubusercontent.com/3384451/89344729-55aa5c80-d674-11ea-8d3b-ca4c409465d4.jpg" />

Something else I noticed this time is in relation to the status bar color. I was able to make the transition to white without reverting back to the blue color
- Open MLP
- Select Create blank page
- Select Up button or back
notice bar is now white

| img | video |
| --- | --- |
| <img width=300 src="https://user-images.githubusercontent.com/3384451/89345105-e6813800-d674-11ea-8cb1-476275e49c08.jpg" /> | <img width=300 src="https://user-images.githubusercontent.com/3384451/89345158-fb5dcb80-d674-11ea-8d61-02a4ac0a13b1.gif" /> |

antonis commented:
Hello @chipsnyder 👋 ,
thank you once again for the feedback 👍 
I fixed the status bar color issue and tweaked the header snap point a bit. Actually it is now a configurable dimension (`mlp_header_scroll_snap_threshold`) we can change based on @iamthomasbishop feedback. 

![device-2020-08-05-093237](https://user-images.githubusercontent.com/304044/89380502-44a72d00-d700-11ea-961a-08ce50f39782.gif)

ashiagr commented:
Hey Antonis 👋

I tested MLP, overall it looks good to me. Couple of observations: 

- On transitioning from landscape to portrait, title was not displayed until I significantly scrolled down (it didn't happen consistently - just once in a while). Let me know if it is an expected behaviour.

![title](https://user-images.githubusercontent.com/1405144/89500015-99b97080-d7de-11ea-8cd7-3b8ae14a04a4.gif)

- Wondering if we could use `CollapsingToolbarLayout` inside a bottom sheet for a smoother title transitioning. Let me know if you already considered it or encountered any challenges. Here's a related [SO discussion](https://rb.gy/brwjzi).

antonis commented:
Hello @ashiagr 👋 ,
thank you for your feedback. I really appreciate it :)

On your observations:

> * On transitioning from landscape to portrait, title was not displayed until I significantly scrolled down (it didn't happen consistently - just once in a while). Let me know if it is an expected behaviour.

I reproduced this behavior and fixed this.

> * Wondering if we could use `CollapsingToolbarLayout` inside a bottom sheet for a smoother title transitioning. Let me know if you already considered it or encountered any challenges. Here's a related [SO discussion](https://rb.gy/brwjzi).

I agree with you on this. An implementation based on `CoordinatorLayout` and `CollapsingToolbarLayout` would be much simpler and cleaner. When I started the implementation I tried this direction but it proved tricky inside the `BottomSheetDialogFragment`. I spend again a few hours on this today since the SO link was quite helpful but I got to a point where I got inconsistent behavior since the `BottomSheetDialogFragment` combined with the nested ([part of a next PR](https://github.com/wordpress-mobile/WordPress-Android/pull/12573)) `RecyclerView`s make this quite complex. I decided that at this point it might be better to proceed with the current solution and revisit this in a next PR as an optimisation effort. Wdyt?

ashiagr commented:
> I reproduced this behavior and fixed this.

Thanks, I confirm it is fixed now 🎉. 

> at this point it might be better to proceed with the current solution and revisit this in a next PR as an optimisation effort. Wdyt?

Thanks for giving it a try 🙇‍♀️. I agree with your suggestion, let's proceed with the current solution. Using `CollapsingToolbarLayout` might not a be a straightforward solution. Whenever (and if) you decide to retry let me know. I'll be happy to pair on this any time after approx. two weeks.

antonis commented:
A new build time configuration `MODAL_LAYOUT_PICKER` enables this feature.
At this point it is only enabled in `wasabi` and `jalapeno` flavors and won't affect alpha, beta and release build
